app/src/main/java/com/keylesspalace/tusky/components/compose/ComposeViewModel.kt

```diff
@@ -121,22 +121,36 @@ class ComposeViewModel @Inject constructor(
         }
     }
 
-    private suspend fun addMediaToQueue(
+    suspend fun addMediaToQueue(
         type: QueuedMedia.Type,
         uri: Uri,
         mediaSize: Long,
-        description: String? = null
+        description: String? = null,
+        replaceItem: QueuedMedia? = null
     ): QueuedMedia {
-        val mediaItem = media.updateAndGet { mediaValue ->
+        var stashMediaItem: QueuedMedia? = null
+
+        media.updateAndGet { mediaValue ->
             val mediaItem = QueuedMedia(
                 localId = (mediaValue.maxOfOrNull { it.localId } ?: 0) + 1,
                 uri = uri,
                 type = type,
                 mediaSize = mediaSize,
                 description = description
             )
-            mediaValue + mediaItem
-        }.last()
+            stashMediaItem = mediaItem
+
+            if (replaceItem != null) {
+                mediaToJob[replaceItem.localId]?.cancel()
+                mediaValue.map {
+                    if (it.localId == replaceItem.localId) mediaItem else it
+                }
+            } else { // Append
+                mediaValue + mediaItem
+            }
+        }
+        val mediaItem = stashMediaItem!! // stashMediaItem is always non-null and uncaptured at this point, but Kotlin doesn't know that
```

Comment:
`if` can return a value, you can do something like

```
val mediaItem = if (...) {
    ...
    returnValueA
} else {
    ...
    returnValueB
}
```

Reply:
Not through the updateAndGet . The value returned by the updateAndGet block is the new value of the List inside "media", so afaict we have to use a var to smuggle a "second" value (like the QueuedMedia) out. But we want to create the new QueuedMedia inside the updateAndGet block so the localId is created atomically with adding the item to the list.